### PR TITLE
CB-8706, CB-8707 saveToPhotoAlbum improvements and some refactoring

### PR DESCRIPTION
**Pull request by muratsu**
- On Windows we should use filePicker instead of programmatically accessing user libs. This way the app doesnt have to depend on extra capabilities.
- Some refactoring to help on navigating code easier

### CODE REVIEW

**Comment by nikhilkh** [comment]
@vladimir-kotikov @sgrebnov Can you please review this?

File: src/windows/CameraProxy.js [hunk 1]
```diff
@@ -43,399 +43,468 @@ module.exports = {
...
-                        });
-                    };
-                });
-            }, function () {
-                errorCallback("Can't access localStorage folder");
-            });
 
-        };
+// Resize method
+function resizeImage(successCallback, errorCallback, file, targetWidth, targetHeight, encodingType) {
```
Comment (vladimir-kotikov) [inline src/windows/CameraProxy.js]: This method contains a lot of duplicated functionality from `resizeImageBase64`. Consider reuse of `resizeImageBase64` for B64 encoding purposes.

File: src/windows/CameraProxy.js [hunk 2]
```diff
@@ -43,399 +43,468 @@ module.exports = {
...
-            // WP8.1 doesn't allow to use of pickSingleFileAsync method
-            // see http://msdn.microsoft.com/en-us/library/windows/apps/br207852.aspx for details
-            // replacement of pickSingleFileAsync - pickSingleFileAndContinue method
-            // will take application to suspended state and this require additional logic to wake application up
-            if (navigator.appVersion.indexOf('Windows Phone 8.1') >= 0 ) {
-                errorCallback('Not supported');
-                return;
-            }
+    fileOpenPicker.pickSingleFileAsync().done(function (file) {
+        if (file) {
```
Comment (vladimir-kotikov) [inline src/windows/CameraProxy.js]: Consider invert `if` statement to remove extra nesting.

File: src/windows/CameraProxy.js [hunk 3]
```diff
@@ -43,399 +43,468 @@ module.exports = {
...
+        captureSettings.streamingCaptureMode = CaptureNS.StreamingCaptureMode.video;
+    };
+
+    var startCameraPreview = function () {
+        // Search for available camera devices
+        // This is necessary to detect which camera (front or back) we should use
+        var expectedPanel = cameraDirection === 1 ? Windows.Devices.Enumeration.Panel.front : Windows.Devices.Enumeration.Panel.back;
+        Windows.Devices.Enumeration.DeviceInformation.findAllAsync(Windows.Devices.Enumeration.DeviceClass.videoCapture)
+        .done(function (devices) {
+            if (devices.length > 0) {
```
Comment (vladimir-kotikov) [inline src/windows/CameraProxy.js]: Consider invert `if` statement to remove extra nesting.

File: src/windows/CameraProxy.js [hunk 4]
```diff
@@ -43,399 +43,468 @@ module.exports = {
...
-                                    }
-                                }
-                            };
-
-                            if (saveToPhotoAlbum) {
-                                capturedFile.copyAsync(Windows.Storage.KnownFolders.picturesLibrary, capturedFile.name, generateUniqueCollisionOption)
-                                .done(function() {
-                                    success(capturedFile);
-                                }, errorCallback);
+                if (saveToPhotoAlbum) {
```
Comment (vladimir-kotikov) [inline src/windows/CameraProxy.js]: Consider invert `if` statement to remove extra nesting.

File: src/windows/CameraProxy.js [hunk 5]
```diff
@@ -43,399 +43,468 @@ module.exports = {
...
+    try {
+        createCameraUI();
+        startCameraPreview();
+    } catch (ex) {
+        errorCallback(ex);
+    }
+}
+
+function takePictureFromCameraWindows(successCallback, errorCallback, args) {
+    var destinationType = args[1],
```
Comment (vladimir-kotikov) [inline src/windows/CameraProxy.js]: nit: There is exactly the same code exist in `takePictureFromCameraWP`. You may want to move it to upper-level `takePictureFromCamera`, construct an options object and pass it downstream.

File: src/windows/CameraProxy.js [hunk 6]
```diff
@@ -43,399 +43,468 @@ module.exports = {
...
+    } else if (targetWidth >= 320 || targetHeight >= 240) {
+        cameraCaptureUI.photoSettings.maxResolution = Windows.Media.Capture.CameraCaptureUIMaxPhotoResolution.verySmallQvga;
+    } else {
+        cameraCaptureUI.photoSettings.maxResolution = Windows.Media.Capture.CameraCaptureUIMaxPhotoResolution.highestAvailable;
+    }
 
-                if (encodingType == Camera.EncodingType.PNG) {
-                    cameraCaptureUI.photoSettings.format = Windows.Media.Capture.CameraCaptureUIPhotoFormat.png;
+    cameraCaptureUI.captureFileAsync(Windows.Media.Capture.CameraCaptureUIMode.photo).done(function(picture) {
+        if (picture) {
```
Comment (vladimir-kotikov) [inline src/windows/CameraProxy.js]: Consider invert `if` statement to remove extra nesting.

File: src/windows/CameraProxy.js [hunk 7]
```diff
@@ -43,399 +43,468 @@ module.exports = {
...
-                    cameraCaptureUI.photoSettings.maxResolution = Windows.Media.Capture.CameraCaptureUIMaxPhotoResolution.large3M;
-                } else if (targetWidth >= 1024 || targetHeight >= 768) {
-                    cameraCaptureUI.photoSettings.maxResolution = Windows.Media.Capture.CameraCaptureUIMaxPhotoResolution.mediumXga;
-                } else if (targetWidth >= 800 || targetHeight >= 600) {
-                    cameraCaptureUI.photoSettings.maxResolution = Windows.Media.Capture.CameraCaptureUIMaxPhotoResolution.mediumXga;
-                } else if (targetWidth >= 640 || targetHeight >= 480) {
-                    cameraCaptureUI.photoSettings.maxResolution = Windows.Media.Capture.CameraCaptureUIMaxPhotoResolution.smallVga;
-                } else if (targetWidth >= 320 || targetHeight >= 240) {
-                    cameraCaptureUI.photoSettings.maxResolution = Windows.Media.Capture.CameraCaptureUIMaxPhotoResolution.verySmallQvga;
+            if (saveToPhotoAlbum) {
```
Comment (vladimir-kotikov) [inline src/windows/CameraProxy.js]: Consider invert `if` statement to remove extra nesting.

File: src/windows/CameraProxy.js [hunk 8]
```diff
@@ -43,399 +43,468 @@ module.exports = {
...
-                var imageData = "data:" + file.contentType + ";base64," + strBase64;
+// Because of asynchronous method, so let the successCallback be called in it.
+function resizeImageBase64(successCallback, errorCallback, file, targetWidth, targetHeight) {
+    Windows.Storage.FileIO.readBufferAsync(file).done( function(buffer) {
+        var strBase64 = Windows.Security.Cryptography.CryptographicBuffer.encodeToBase64String(buffer);
+        var imageData = "data:" + file.contentType + ";base64," + strBase64;
 
-                var image = new Image();
-                image.src = imageData;
+        var image = new Image();
```
Comment (vladimir-kotikov) [inline src/windows/CameraProxy.js]: Not sure if this is in scope of this PR but we may want to rewrite resize functionality using classes from [Windows.Graphics.Imaging namespace](https://msdn.microsoft.com/en-us/library/windows/apps/br211940.aspx).

File: src/windows/CameraProxy.js (hunk 1 above)
Comment (muratsu) [inline src/windows/CameraProxy.js]: I'll make this change in a separate commit I think.

File: src/windows/CameraProxy.js (hunk 7 above)
Comment (muratsu) [inline src/windows/CameraProxy.js]: fixed

File: src/windows/CameraProxy.js (hunk 6 above)
Comment (muratsu) [inline src/windows/CameraProxy.js]: fixed

File: src/windows/CameraProxy.js (hunk 5 above)
Comment (muratsu) [inline src/windows/CameraProxy.js]: i'll address this in a separate PR

File: src/windows/CameraProxy.js (hunk 8 above)
Comment (muratsu) [inline src/windows/CameraProxy.js]: i'll address this in a separate PR

File: src/windows/CameraProxy.js (hunk 2 above)
Comment (muratsu) [inline src/windows/CameraProxy.js]: fixed

File: src/windows/CameraProxy.js (hunk 3 above)
Comment (muratsu) [inline src/windows/CameraProxy.js]: fixed

File: src/windows/CameraProxy.js (hunk 4 above)
Comment (muratsu) [inline src/windows/CameraProxy.js]: fixed

**Comment by muratsu** [comment]
@vladimir-kotikov @sgrebnov ping

**Comment by vladimir-kotikov** [comment]
LGTM

**Pull request merged**